multistream-select/src/length_delimited.rs

```diff
@@ -0,0 +1,319 @@
+    // of data itself before it is returned.
+    // Must always contain enough space to read data from `inner`.
+    internal_buffer: SmallVec<[u8; 64]>,
+    // Number of bytes within `internal_buffer` that contain valid data.
+    internal_buffer_pos: usize,
+    // State of the decoder.
+    state: State,
+    marker: PhantomData<I>,
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+enum State {
+    // We are currently reading the length of the next frame of data.
+    ReadingLength,
+    // We are currently reading the frame of data itself.
+    ReadingData {
+        frame_len: u16
+    },
+}
+
+impl<I, S> LengthDelimitedFramedRead<I, S> {
+    pub fn new(inner: S) -> LengthDelimitedFramedRead<I, S> {
+        LengthDelimitedFramedRead {
+            inner: inner,
+            internal_buffer: { let mut v = SmallVec::new(); v.push(0); v },
+            internal_buffer_pos: 0,
+            state: State::ReadingLength,
+            marker: PhantomData,
+        }
+    }
+
+    /// Destroys the `LengthDelimitedFramedRead` and returns the underlying socket.
+    ///
+    /// Contrary to its equivalent `tokio_io::codec::length_delimited::FramedRead`, this method is
+    /// guaranteed not to skip any data from the socket.
+    ///
+    /// # Panic
+    ///
+    /// Will panic if called while there is data inside the buffer. **This can only happen if
+    /// you call `poll()` manually**. Using this struct as it is intended to be used (ie. through
+    /// the modifiers provided by the `futures` crate) will always leave the object in a state in
+    /// which `into_inner()` will not panic.
+    #[inline]
+    pub fn into_inner(self) -> S {
+        assert_eq!(self.state, State::ReadingLength);
+        assert_eq!(self.internal_buffer_pos, 0);
+        self.inner
+    }
+}
+
+impl<I, S> Stream for LengthDelimitedFramedRead<I, S>
+    where S: AsyncRead,
+          I: for<'r> From<&'r [u8]>
+{
+    type Item = I;
+    type Error = IoError;
+
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        loop {
+            debug_assert!(!self.internal_buffer.is_empty());
+            debug_assert!(self.internal_buffer_pos < self.internal_buffer.len());
+
+            match self.state {
+                State::ReadingLength => {
+                    match self.inner.read(&mut self.internal_buffer[self.internal_buffer_pos..]) {
+                        Ok(0) => {
+                            // EOF
+                            if self.internal_buffer_pos == 0 {
+                                return Ok(Async::Ready(None));
+                            } else {
+                                return Err(IoError::new(IoErrorKind::BrokenPipe,
+                                                        "unexpected eof"));
+                            }
+                        },
+                        Ok(n) => {
+                            debug_assert_eq!(n, 1);
```

Comment:
why is it always 1?

Reply:
By design, when reading the length prefix we read one byte at a time (see lines 68, 148 and 182). The length prefix itself is only one byte long 99.9% of the time, and we only accept a maximum of 2 bytes, so I don't think it's a problem.

Reply:
I see it now, thanks!